File: src/main/java/com/beat/global/auth/jwt/filter/JwtAuthenticationFilter.java
package com.beat.global.auth.jwt.filter;

import com.beat.domain.user.domain.Role;
import com.beat.global.auth.jwt.provider.JwtTokenProvider;
import com.beat.global.auth.jwt.provider.JwtValidationType;
import com.beat.global.auth.security.AdminAuthentication;
import com.beat.global.auth.security.MemberAuthentication;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

import java.io.IOException;
import java.util.Collection;
import java.util.List;

import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;
import org.springframework.web.filter.OncePerRequestFilter;

@Slf4j
@Component
@RequiredArgsConstructor
public class JwtAuthenticationFilter extends OncePerRequestFilter {

	private final JwtTokenProvider jwtTokenProvider;

	@Override
	protected void doFilterInternal(@NonNull HttpServletRequest request,
		@NonNull HttpServletResponse response,
		@NonNull FilterChain filterChain) throws ServletException, IOException {

		final String token = getJwtFromRequest(request);

		if (!StringUtils.hasText(token)) {
			log.info("JWT Token not found in request header. Assuming guest access or public API request.");
			filterChain.doFilter(request, response);
			return;
		}

		try {
			JwtValidationType validationType = jwtTokenProvider.validateToken(token);

			if (validationType == JwtValidationType.VALID_JWT) {
				setAuthentication(token, request);
				filterChain.doFilter(request, response);
			} else {
				handleInvalidToken(validationType, response);
			}
		} catch (Exception e) {
			log.error("JWT Authentication Exception: ", e);
			response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR); // 500 응답
		}
	}

	private void setAuthentication(String token, HttpServletRequest request) {
		Long memberId = jwtTokenProvider.getMemberIdFromJwt(token);
		Role role = jwtTokenProvider.getRoleFromJwt(token);

		log.info("Setting authentication for memberId: {} with role: {}", memberId, role);

		Collection<GrantedAuthority> authorities = List.of(role.toGrantedAuthority());
		UsernamePasswordAuthenticationToken authentication = createAuthentication(memberId, authorities, role);
		authentication.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));

		log.info("Authentication set: memberId: {}, role: {}", memberId, role);

		SecurityContextHolder.getContext().setAuthentication(authentication);
	}

	private void handleInvalidToken(JwtValidationType validationType, HttpServletResponse response) {
		if (validationType == JwtValidationType.EXPIRED_JWT_TOKEN) {
			log.warn("JWT Token is expired");
			response.setStatus(HttpServletResponse.SC_UNAUTHORIZED); // 401 응답
		} else {
			log.warn("JWT Token is invalid");
			response.setStatus(HttpServletResponse.SC_BAD_REQUEST); // 400 응답
		}
	}

	private UsernamePasswordAuthenticationToken createAuthentication(Long memberId,
		Collection<GrantedAuthority> authorities, Role role) {
		log.info("Creating authentication for memberId: {} with role: {}", memberId, role);

		if (role == Role.ADMIN) {
			log.info("Creating AdminAuthentication for memberId: {}", memberId);
			return new AdminAuthentication(memberId.toString(), null, authorities);
		} else if (role == Role.MEMBER) {
			log.info("Creating MemberAuthentication for memberId: {}", memberId);
			return new MemberAuthentication(memberId.toString(), null, authorities);
		}
		log.error("Unknown role: {}", role);
		throw new IllegalArgumentException("Unknown role: " + role);
	}

	private String getJwtFromRequest(HttpServletRequest request) {
		String bearerToken = request.getHeader("Authorization");
		if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
			return bearerToken.substring("Bearer ".length());
		}
		return null;
	}
}

File: support/security-web/src/main/kotlin/com/beat/support/security/authentication/internal/JwtAuthenticationFilter.kt
package com.beat.support.security.authentication.internal

import com.beat.support.observability.logging.filter.BaseMdcLoggingFilter
import com.beat.support.security.access.AccessTokenAuthenticationFailure
import com.beat.support.security.access.AccessTokenAuthenticationResult
import com.beat.support.security.access.AccessTokenAuthenticator
import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.MDC
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component("gatewayJwtAuthenticationFilter")
internal class JwtAuthenticationFilter(
    private val accessTokenAuthenticator: AccessTokenAuthenticator,
) : OncePerRequestFilter() {

    private val authenticationDetailsSource = WebAuthenticationDetailsSource()

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        val token = request.bearerToken()

        if (token == null) {
            filterChain.doFilter(request, response)
            return
        }

        try {
            when (val result = accessTokenAuthenticator.authenticateAccessToken(token)) {
                is AccessTokenAuthenticationResult.Authenticated -> {
                    authenticate(result.memberId, result.roleName, request)
                    filterChain.doFilter(request, response)
                }

                is AccessTokenAuthenticationResult.Rejected -> {
                    response.status = result.failure.toHttpStatus()
                }
            }
        } catch (_: IllegalArgumentException) {
            response.status = HttpServletResponse.SC_UNAUTHORIZED
        } catch (exception: Exception) {
            log.error(exception) { "Unexpected JWT authentication failure" }
            response.status = HttpServletResponse.SC_INTERNAL_SERVER_ERROR
        }
    }

    private fun authenticate(memberId: Long, roleName: String, request: HttpServletRequest) {
        val authentication = createAuthentication(memberId, roleName).apply {
            details = authenticationDetailsSource.buildDetails(request)
        }

        SecurityContextHolder.getContext().authentication = authentication
        MDC.put(BaseMdcLoggingFilter.USER_ID_KEY, memberId.toString())
    }

    private fun createAuthentication(memberId: Long, roleName: String): UsernamePasswordAuthenticationToken {
        val authorities: Collection<GrantedAuthority> = listOf(SimpleGrantedAuthority(roleName))
        return when (roleName) {
            ROLE_ADMIN -> AdminAuthentication(memberId, authorities)
            ROLE_MEMBER -> MemberAuthentication(memberId, authorities)
            else -> throw IllegalArgumentException("Unsupported role: $roleName")
        }
    }

    /**
     * `EXPIRED -> 401`, 그 외 실패 -> `400`은 현재 client 호환 계약이다.
     */
    private fun AccessTokenAuthenticationFailure.toHttpStatus(): Int = when (this) {
        AccessTokenAuthenticationFailure.EXPIRED -> HttpServletResponse.SC_UNAUTHORIZED
        else -> HttpServletResponse.SC_BAD_REQUEST
    }

    private fun HttpServletRequest.bearerToken(): String? =
        getHeader(HEADER_AUTHORIZATION)
            ?.takeIf { it.startsWith(BEARER_PREFIX) }
            ?.removePrefix(BEARER_PREFIX)
            ?.takeIf(String::isNotBlank)

    companion object {
        private val log = KotlinLogging.logger {}
        private const val ROLE_ADMIN = "ROLE_ADMIN"
        private const val ROLE_MEMBER = "ROLE_MEMBER"
        private const val HEADER_AUTHORIZATION = "Authorization"
        private const val BEARER_PREFIX = "Bearer "
    }
}
